File: src/p1008/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode *bstFromPreorder(const vector<int> &v) {
        i = 0; return build(v, INT_MAX);
    }

private:
    size_t i;
    TreeNode *build(const vector<int> &v, int bound) {
        if (i < v.size() && v[i] <= bound) {
            auto node = new TreeNode(v[i++]);
            node->left = build(v, node->val);
            node->right = build(v, bound);
            return node;
        }
        return nullptr;
    }
};

File: src/p1008/java/solution.java
class Solution {
    public TreeNode bstFromPreorder(int[] v) {
        i = 0;
        return build(v, Integer.MAX_VALUE);
    }
    private int i;
    private TreeNode build(int[] v, int bound) {
        if (i < v.length && v[i] <= bound) {
            TreeNode node = new TreeNode(v[i++]);
            node.left = build(v, node.val);
            node.right = build(v, bound);
            return node;
        }
        return null;
    }
}
